Sources/Basics/Observability.swift

```diff
@@ -0,0 +1,400 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import TSCBasic
+import TSCUtility
+
+typealias TSCDiagnostic = TSCBasic.Diagnostic
+
+// this could become a struct when we remove the "errorsReported" pattern
+
+// designed after https://github.com/apple/swift-log
+// designed after https://github.com/apple/swift-metrics
+public class ObservabilitySystem {
+
+    // global
+
+    private static var _global = ObservabilitySystem(factory: NOOPFactory())
+    private static var bootstrapped = false
+    private static let lock = Lock()
+
+    public static func bootstrapGlobal(factory: ObservabilityFactory) {
+        Self.lock.withLock {
+            // FIXME: disabled for testing
+            //precondition(!Self.bootstrapped, "ObservabilitySystem can only bootstrapped once")
+            Self._global = .init(factory: factory)
+            Self.bootstrapped = true
+        }
+    }
+
+    @available(*, deprecated, message: "this pattern is deprecated, transition to error handling instead")
+    public static var errorsReported: Bool {
+        Self.lock.withLock {
+            Self._global.errorsReported
+        }
+    }
+
+    // as we transition to async/await we can take advantage of Task Local values instead of a global
+    fileprivate static var global: ObservabilitySystem {
+        Self.lock.withLock {
+            Self._global
+        }
+    }
+
+    // compatibility with DiagnosticsEngine
+
+    @available(*, deprecated, message: "temporary for transition DiagnosticsEngine -> DiagnosticsEmitter")
+    public static func makeDiagnosticsEngine() -> DiagnosticsEngine {
+        return DiagnosticsEngine(handlers: [{ Diagnostic($0).map{ self.global.diagnosticsHandler($0) } }])
+    }
+
+    // instance
+
+    public var diagnosticsHandler: DiagnosticsHandler!
+    private var _errorsReported = ThreadSafeBox<Bool>(false)
+
+    public init(factory: ObservabilityFactory) {
+        self.diagnosticsHandler = { diagnostic in
+            if diagnostic.severity == .error {
+                self._errorsReported.put(true)
+            }
+            factory.diagnosticsHandler(diagnostic)
+        }
+    }
+
+    // FIXME: we want to remove this functionality and move to more conventional error handling
```

Comment:
Does that refer to throwing `Swift.Error`?  If so, the main issue with that is that it's not apparent how to report more than a single error at once.  Are you thinking that a thrown `Swift.Error` could carry multiple diagnostics?  Or does this conventional error handling refer to something other than throwing `Swift.Error`s?

Reply:
we have a pattern for this in the collections area of the code, which I was thinking to pull into here too. but this is all for later PR dedicated to that

Reply:
Sounds great.